bfabric/src/bfabric/entities/core/has_many.py

```diff
@@ -15,89 +15,52 @@
 T = TypeVar("T")
 
 
-class HasMany(Relationship[E]):
+class HasMany(Generic[E]):
     def __init__(
         self,
-        entity: str,
+        entity: None = None,
         *,
         bfabric_field: str | None = None,
-        ids_property: str | None = None,
-        client_property: str = "_client",
         optional: bool = False,
     ) -> None:
-        super().__init__(entity)
         self._bfabric_field = bfabric_field
-        self._ids_property = ids_property
-        self._client_property = client_property
         self._optional = optional
 
-    def __get__(self, obj: T | None, objtype: type[T] | None = None) -> _HasManyProxy:
-        cache_attr = f"_HasMany__{self._ids_property or self._bfabric_field}_cache"
-        if obj is None:
-            raise ValueError("Cannot access HasMany relationship on class")
-        if not hasattr(obj, cache_attr):
-            ids = self._get_ids(obj)
-            client = getattr(obj, self._client_property)
-            setattr(
-                obj,
-                cache_attr,
-                _HasManyProxy(entity_type=self._entity_type, ids=ids, client=client),
-            )
-        return getattr(obj, cache_attr)
-
-    def _get_ids(self, obj: T) -> list[int]:
-        if self._bfabric_field is not None:
-            if self._ids_property is not None:
-                raise ValueError("Exactly one of bfabric_field and ids_property must be set, but both are set")
-            if self._optional and self._bfabric_field not in obj.data_dict:
-                return []
-            return [x["id"] for x in obj.data_dict[self._bfabric_field]]
-        elif self._ids_property is not None:
-            if self._optional and not hasattr(obj, self._ids_property):
-                return []
-            return getattr(obj, self._ids_property)
-        else:
-            raise ValueError("Exactly one of bfabric_field and ids_property must be set, but neither is set")
+    def __get__(self, obj: T | None, objtype: type[T] | None = None) -> _HasManyProxy[E]:
+        items = obj.refs.get(self._bfabric_field)
+        if items is None and not self._optional:
+            raise ValueError(f"Missing field: {self._bfabric_field}")
+        items = items or []
+        return _HasManyProxy(items=items)
 
 
 class _HasManyProxy(Generic[E]):
-    def __init__(self, entity_type: type[E], ids: list[int], client: Bfabric) -> None:
-        self._entity_type = entity_type
-        self._ids = ids
-        self._client = client
-        self._items: dict[int, E] = {}
+    def __init__(self, items: list[E]) -> None:
+        self._items = items
 
     @property
     def ids(self) -> list[int]:
-        return self._ids
+        return [item.id for item in self._items]
```

Comment:
This drops the optimization (which I'm not sure if it worked in the first place) to not fetch the items when all you need is the ids.

The new API for this purpose is `References.uris` which should provide all the necessary information without performing any requests.

Reply:
Maybe this should be split off into a new issue to figure out where it is used and should be handled differently